src/cares_wrap.h

```diff
@@ -0,0 +1,639 @@
+      idx = 0;
+      while (host->h_addr_list[idx]) {
+        free(host->h_addr_list[idx++]);
+      }
+      free(host->h_addr_list);
+      host->h_addr_list = nullptr;
+    }
+
+    if (host->h_aliases != nullptr) {
+      idx = 0;
+      while (host->h_aliases[idx]) {
+        free(host->h_aliases[idx++]);
+      }
+      free(host->h_aliases);
+      host->h_aliases = nullptr;
+    }
+
+    free(host->h_name);
+    free(host);
+  }
+
+  using HostEntPointer = DeleteFnPtr<hostent, safe_free_hostent>;
+
+  struct ResponseData {
+    int status;
+    bool is_host;
+    HostEntPointer host;
+    MallocedBuffer<unsigned char> buf;
+  };
+
+  void AfterResponse() {
+    CHECK(response_data_);
+
+    const int status = response_data_->status;
+
+    if (status != ARES_SUCCESS) {
+      ParseError(status);
+    } else if (!response_data_->is_host) {
+      Parse(response_data_->buf.data, response_data_->buf.size);
+    } else {
+      Parse(response_data_->host.get());
+    }
+  }
+
+  void* MakeCallbackPointer() {
+    CHECK_NULL(callback_ptr_);
+    callback_ptr_ = new QueryWrap<Traits>*(this);
+    return callback_ptr_;
+  }
+
+  static QueryWrap<Traits>* FromCallbackPointer(void* arg) {
+    std::unique_ptr<QueryWrap<Traits>*> wrap_ptr {
+        static_cast<QueryWrap<Traits>**>(arg)
+    };
+    QueryWrap<Traits>* wrap = *wrap_ptr.get();
+    if (wrap == nullptr) return nullptr;
+    wrap->callback_ptr_ = nullptr;
+    return wrap;
+  }
+
+  static void Callback(
+      void* arg,
+      int status,
+      int timeouts,
+      unsigned char* answer_buf,
+      int answer_len) {
+    QueryWrap<Traits>* wrap = FromCallbackPointer(arg);
+    if (wrap == nullptr) return;
+
+    unsigned char* buf_copy = nullptr;
+    if (status == ARES_SUCCESS) {
+      buf_copy = node::Malloc<unsigned char>(answer_len);
+      memcpy(buf_copy, answer_buf, answer_len);
+    }
+
+    wrap->response_data_ = std::make_unique<ResponseData>();
+    ResponseData* data = wrap->response_data_.get();
+    data->status = status;
+    data->is_host = false;
+    data->buf = MallocedBuffer<unsigned char>(buf_copy, answer_len);
+
+    wrap->QueueResponseCallback(status);
+  }
+
+  static void Callback(
+      void* arg,
+      int status,
+      int timeouts,
+      struct hostent* host) {
+    QueryWrap<Traits>* wrap = FromCallbackPointer(arg);
+    if (wrap == nullptr) return;
+
+    struct hostent* host_copy = nullptr;
+    if (status == ARES_SUCCESS) {
+      host_copy = node::Malloc<hostent>(1);
+      cares_wrap_hostent_cpy(host_copy, host);
+    }
+
+    wrap->response_data_ = std::make_unique<ResponseData>();
+    ResponseData* data = wrap->response_data_.get();
+    data->status = status;
+    data->host.reset(host_copy);
+    data->is_host = true;
+
+    wrap->QueueResponseCallback(status);
+  }
+
+  void QueueResponseCallback(int status) {
+    BaseObjectPtr<QueryWrap<Traits>> strong_ref{this};
+    env()->SetImmediate([this, strong_ref](Environment*) {
```

Comment:
Are we doing some fake-async work? Why the `SetImmediate`?

Reply:
git blame has the answer: 5a948f6f64a538dedce969b21cde454600197ded

tl;dr: if you call `dns.setServers()` from a synchronous callback, the process is going to crash.

Reply:
Amazing work! We should really add a comment here.

Reply:
This is one bit that I've never really been happy with. Rather than introducing the `SetImmediate()` call here, we really ought to block the ability to call `dns.setServers()` from inside a synchronous callback. Calling `setServers()` is unlikely ever to be a hotpath, while the cost of the `SetImmediate()` is incurred on every query. Unfortunately, changing that behavior would be semver-major and is out of scope for this particular PR.

Reply:
I think this is fine, tbh. `SetImmediate()` is a fairly low-overhead operation, compared to the actual I/O.

Reply:
Either way, I don't plan on addressing that in this PR.